Write a tutorial for dates, schedules, calendars and dcf (#468)
digital-asset/docs.daml.com: docs/2.8.0/docs/daml-finance/tutorials/advanced-topics/instrument-modeling/calendar-dcf.rst

.. Copyright (c) 2023 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.

> .. SPDX-License-Identifier: Apache-2.0

Reviewer: rename the file to `date-utilities`?

Author: After I changed the name, GitHub would not let me Commit your suggestions. Applied them manually instead.

How to use the calendar, schedule and day count utility functions
#################################################################

The Daml Finance library contains date related utility functions that are used to implement industry
standard conventions. These are mainly used internally in instruments like Bonds and Swaps. However,

> there are use cases where it would be helfpul for users to understand these utility functions in

Reviewer: ```suggestion
there are use cases where it would be helpful for users to understand these utility functions in
```

> more detail, for example if you want to:

Reviewer: ```suggestion
more detail. For example if you want to:
```

- develop your own instruments that depend on business day shifts, schedules, and/or day count
  conventions

> - understand and validate the implementation of current Daml Finance instruments

Reviewer: ```suggestion
- understand and validate the implementation of existing Daml Finance instruments
```

- compose existing Daml Finance instruments, for example define a series of zero coupon bonds
  according to a specific schedule of expiries

> - use this calendar and schedule related functionality for non-financial use cases

Reviewer: ```suggestion
- use these calendar and schedule related functionalities for non-financial use cases
```

Calendar
========

The :ref:`Calendar <module-daml-finance-util-date-calendar-17588>` module contains various utility
functions related to business days and date adjustments.

Business days and non-business days are distinguished by defining a holiday calendar, for example:

.. literalinclude:: ../../../src/test/daml/Daml/Finance/Util/Test/Date/Calendar.daml
  :language: daml
  :start-after: -- CREATE_HOLIDAY_CALENDAR_DATA_BEGIN
  :end-before: -- CREATE_HOLIDAY_CALENDAR_DATA_END

The :ref:`HolidayCalendarData <type-daml-finance-interface-types-date-calendar-holidaycalendardata-60004>`
type describes the set of non-business days by defining which days of the week belong to the weekend

> and which dates are general holidays.

Reviewer: ```suggestion
and which dates are additional holidays.
```

  :language: daml
  :start-after: -- TEST_ADD_BUSINESS_DAYS_NEGATIVE_OFFSET_BEGIN
  :end-before: -- TEST_ADD_BUSINESS_DAYS_NEGATIVE_OFFSET_END

Finally, it is also possible to adjust a date according to a given business day convention:

.. literalinclude:: ../../../src/test/daml/Daml/Finance/Util/Test/Date/Calendar.daml
  :language: daml
  :start-after: -- TEST_ADJUST_DATE_BEGIN
  :end-before: -- TEST_ADJUST_DATE_END

The different conventions are described in the
:ref:`BusinessDayConventionEnum <type-daml-finance-interface-types-date-calendar-businessdayconventionenum-88986>`.
In particular, they describe what should happen if the next/previous business day is around month
end.

RollConvention
==============

The :ref:`RollConvention <module-daml-finance-util-date-rollconvention-88672>` module provides
utility functions to add a period to a given date.

For example, you can add a day, week, month or a year to a given date:

.. literalinclude:: ../../../src/test/daml/Daml/Finance/Util/Test/Date/Calendar.daml
  :language: daml
  :start-after: -- TEST_ADD_PERIOD_DAY_WEEK_MONTH_YEAR_BEGIN
  :end-before: -- TEST_ADD_PERIOD_DAY_WEEK_MONTH_YEAR_END

> This function can also handle edge cases, for example in case of a monthly roll to a date that may

Reviewer: We haven't defined what a "roll" is

Alternatively, we can also define a roll to a specific day of the month:

.. literalinclude:: ../../../src/test/daml/Daml/Finance/Util/Test/Date/Calendar.daml
  :language: daml
  :start-after: -- TEST_NEXT_DOM1_BEGIN
  :end-before: -- TEST_NEXT_DOM1_END

If the resulting month does not have enough days, it will default to the last day of the month:

.. literalinclude:: ../../../src/test/daml/Daml/Finance/Util/Test/Date/Calendar.daml
  :language: daml
  :start-after: -- TEST_NEXT_DOM_PAST_EOM_BEGIN
  :end-before: -- TEST_NEXT_DOM_PAST_EOM_END

This function also takes leap days into account:

.. literalinclude:: ../../../src/test/daml/Daml/Finance/Util/Test/Date/Calendar.daml
  :language: daml
  :start-after: -- TEST_NEXT_DOM_PAST_EOFEB_BEGIN
  :end-before: -- TEST_NEXT_DOM_PAST_EOFEB_END

Similarly, it is possible to roll to the *previous* period:

.. literalinclude:: ../../../src/test/daml/Daml/Finance/Util/Test/Date/Calendar.daml
  :language: daml
  :start-after: -- TEST_PREVIOUS_DOM_FROM_EOFEB_BEGIN
  :end-before: -- TEST_PREVIOUS_DOM_FROM_EOFEB_END

Schedule
========

While the above :ref:`RollConvention <module-daml-finance-util-date-rollconvention-88672>` functions

> can be useful on their own, they are often required in the context of rolling out a schedule. This

Reviewer: I think we should define what a schedule is

Reviewer: Might be worth explaining unadjusted vs adjusted dates

.. literalinclude:: ../../../src/test/daml/Daml/Finance/Util/Test/Date/Calendar.daml
  :language: daml
  :start-after: -- CREATE_EXPECTED_SCHEDULE_RESULT_WITH_STUB_BEGIN
  :end-before: -- CREATE_EXPECTED_SCHEDULE_RESULT_WITH_STUB_END

Different stub types can be configured using the
:ref:`StubPeriodTypeEnum <type-daml-finance-interface-types-date-schedule-stubperiodtypeenum-69372>`
data type. You can configure whether the stub period should be at the beginning or at the end of the
schedule, as well as whether the stub period should be longer or shorter than a regular period.

Let us roll out this schedule, which includes a stub period, and compare it to the expected result:

.. literalinclude:: ../../../src/test/daml/Daml/Finance/Util/Test/Date/Calendar.daml
  :language: daml
  :start-after: -- CREATE_SCHEDULE_WITH_STUB_BEGIN
  :end-before: -- CREATE_SCHEDULE_WITH_STUB_END

The schedule can be used to represent different aspects of a financial instrument, for example:

- payment periods, which define e.g. *when* a bond coupon or a swap payment should be paid
- calculation periods, which define e.g. *how* the interest amount is calculated

DayCount
========

> Many instruments that pay interest require an exact definition of how many days of interest belong

Reviewer: We should specify that this is relevant as the coupon rate is often expressed in annualized terms

to each payment period. The :ref:`DayCount <module-daml-finance-util-date-daycount-38239>` module
provides functions for this. In particular, a *day count fraction* (or *dcf* in short) can be
calculated between two dates. This indicates the fraction of a full year.

For example, consider a one year bond that pays a quarterly coupon according to a given schedule,

> like the one described in the previous chapter.

Reviewer: ```suggestion
like the one described in the previous section.
```
